waiter/src/waiter/descriptor.clj

```diff
@@ -0,0 +1,231 @@
+  [handler request->descriptor-fn]
+  (fn [request]
+    (tl/try-let [request-descriptor (request->descriptor-fn request)]
+                (let [{:keys [descriptor latest-service-id]} request-descriptor
+                      handler (middleware/wrap-merge handler {:descriptor descriptor :latest-service-id latest-service-id})]
+                  (handler request))
+                (catch Exception e
+                  (if (missing-run-as-user? e)
+                    (let [{:keys [query-string uri]} request
+                          location (str "/waiter-consent" uri (when (not (str/blank? query-string)) (str "?" query-string)))]
+                      (counters/inc! (metrics/waiter-counter "auto-run-as-requester" "redirect"))
+                      (meters/mark! (metrics/waiter-meter "auto-run-as-requester" "redirect"))
+                      {:headers {"location" location} :status 303})
+                    (do
+                      ; For consistency with historical data, count errors looking up the descriptor as a "process error"
+                      (meters/mark! (metrics/waiter-meter "core" "process-errors"))
+                      (utils/exception->response e request)))))))
+
+(defn fallback-maintainer
+  "Long running daemon process that listens for scheduler state updates and triggers changes in the
+   fallback state. It also responds to queries for the fallback state."
+  [scheduler-state-chan fallback-state-atom]
+  (let [exit-chan (au/latest-chan)
+        query-chan (async/chan 32)
+        channels [exit-chan scheduler-state-chan query-chan]]
+    (async/go
+      (loop [{:keys [available-service-ids healthy-service-ids]
+              :or {available-service-ids #{}
+                   healthy-service-ids #{}} :as current-state}
+             @fallback-state-atom]
+        (let [next-state
+              (try
+                (let [[message selected-chan] (async/alts! channels :priority true)]
+                  (condp = selected-chan
+                    exit-chan
+                    (if (= :exit message)
+                      (log/warn "stopping fallback-maintainer")
+                      (do
+                        (log/info "received unknown message, not stopping fallback-maintainer" {:message message})
+                        current-state))
+
+                    query-chan
+                    (let [{:keys [response-chan service-id]} message]
+                      (->> (if service-id
+                             {:available (contains? available-service-ids service-id)
+                              :healthy (contains? healthy-service-ids service-id)
+                              :service-id service-id}
+                             {:state current-state})
+                           (async/>! response-chan))
+                      current-state)
+
+                    scheduler-state-chan
+                    (let [{:keys [available-service-ids healthy-service-ids]}
```

Comment:
It is okay to look for healthy services from here since the scheduler syncer trigger waiter-specific health checks too. For details please see `scheduler/update-scheduler-state`